game/script-farewells.rpy

```diff
@@ -2,42 +2,757 @@ default persistent._farewell_database = dict()
 
 init python in farewells:
     import random
+    import store
+
+    # Natsuki will not ask a player to stay again if they agreed previously
+    store.jn_globals.player_already_stayed = False
+
     FAREWELL_MAP = dict()
 
+    def get_farewell_in_affinity_range(farewell):
+        """
+        Returns true if player's persistent value is within the farewell topic's affinity range
+
+        IN:
+            farewell - The farewell topic to check
+        """
+        if store.persistent.affinity in range(farewell.affinity_range[0], farewell.affinity_range[1]):
+            return True
+            
+        else:
+            return False
+
+    def get_farewell_has_no_stay_option(farewell):
+        """
+        Returns true if the farewell topic given has no has_stay_option attribute set,
+        or if has_stay_option is set to False
+
+        IN:
+            farewell - The farewell topic to check
+        """
+        if not farewell.additional_properties["has_stay_option"]:
+            return True
+
+        else:
+            return False
+
     def select_farewell():
         """
-        Picks a random greeting
+        Picks a random farewell, accounting for affinity.
+        If the player has already been asked to stay by Natsuki, a farewell without the option
+        to stay will be selected
         """
-        #TODO: ME (for now this just returns a random)
-        return random.choice(FAREWELL_MAP.keys())
+        farewells_in_affinity_range = filter(get_farewell_in_affinity_range, FAREWELL_MAP.values())
```

Comment:
As mentioned in a prior point, we can use a generalized staticmethod on Topic to filter for appropriate events (i.e. passing conditionals + met affection/trust criteria (and eventually start/end date ranges))